README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

# Animated FAB Menu
**Animated FAB Menu** package lets you add an animated floating action button to your Flutter app that open a list of menu items.


## Features
The FabMenu widget is built to be placed in the `Scaffold.floatingActionButton` argument, replacing the `FloatingActionButton` widget. Its position is set by the `fabAlignment` property so you don't need to set it with the `Scaffold.floatingActionButtonLocation` argument. By default the position is set to `bottomCenter`, but you can change it by the `fabAlignment` property to place the button anywhere in the screen. By using the `fabIcon` and `fabBackgroundColor`properties, you can change the icon and the color of the button to whatever you like. 

* *Menu*

By pressing the floating action button, a page containing a **menu** list will be opening with some cool animation. This **menu** contains items that you can specify by the `children` list property. Every child inside the list takes a `title` for the item name, `style` for the text widget of the name, `onTap` function to provide the logic of the on tap action on each child. On top of that, you have an `overlayColor` property to change the color of the background opened page, `overlayOpacity` to change the opacity of the background color and `closeMenuButton` to choose your own back button widget.

*The package will handle the animation by itself.*

![animated_fab_menu](https://user-images.githubusercontent.com/68671238/134653636-1e0f070c-588a-4189-aa7f-45e9957ccae9.png)
  
<hr>

## Getting started

1. Add the latest version of package to your `pubspec.yaml` (and run `dart pub get`):
```dart
dependencies:
  animated_fab_button_menu: ^0.0.2
```
2. Import the package and use it in your Flutter App.
```dart
import 'package:animated_fab_button_menu/animated_fab_button_menu.dart';
```
<hr>

## Usage

There are a number of properties that you can modify:

* fabIcon
* fabBackgroundColor
* fabAlignment
* overlayColor (Color of the background menu)
* overlayOpacity (Opacity of the background menu color)
* elevation
* closeMenuButton (Button that close the menu item)
* children (Items of the menu)
* title (Name of the child menu) 
* style (Text style for the child menu)
* onTap (onTap function for the child menu)

**Example Usage ( complete with all params ):**
<table>
 <tr>
 <td>
      
```dart
class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButton: FabMenu(
          fabBackgroundColor: Colors.amber,
          elevation: 2.0,
          fabAlignment: Alignment.bottomCenter,
          fabIcon: const Icon(Icons.more_horiz),
          closeMenuButton: const Icon(
            Icons.arrow_back,
            color: Colors.white,
          ),
          overlayOpacity: 0.5,
          overlayColor: Colors.amber,
          children: [
            MenuItem(
              title: 'Home',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Profile',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Messages',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Feedback',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Settings',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Share',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Get in touch!',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
          ]),
      body: const Center(
        child: Text('Flutter animated FAB Menu'),
      ),
    );
  }
}
```
   </td>
   <td>
     Here's what it looks like:
     
<hr>


https://user-images.githubusercontent.com/68671238/134640734-751456fa-686b-4e77-bc8d-04ba3a324baf.mp4


   </td>
  </tr>
  </table>
<hr>

## Next Goals
We are working on some improvements including:

- [x] Change color and opacity for menu background. Change color and opacity by specifying `overlayColor` and `overlayOpacity` properties.
- [ ] Make the distance between menu items customizable.
- [ ] Change opening animation with more smooth and user friendly one.

## Issues & Feedback
Please file an [issue!](https://github.com/aliMissaoui/Flutter-Animated-FAB-Menu/issues) to send feedback or report a bug. Thank you!



CHANGELOG.md
## 0.0.1

* Customize and use an animated FAB menu to display an app menu.

## 0.0.2

* beta version + example.


exemple/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:exemple/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/src/fab_menu.dart
import 'package:animated_fab_button_menu/src/item_child.dart';
import 'package:animated_fab_button_menu/src/fab_menu_item.dart';
import 'package:animated_fab_button_menu/src/hero_dialog_route.dart';
import 'package:flutter/material.dart';

class FabMenu extends StatefulWidget {
  /// Used to set the fab Icon.
  final Widget? fabIcon;

  /// Used to set the fab background color.
  final Color? fabBackgroundColor;

  /// Used to set the menu close Icon.
  final Widget? closeMenuButton;

  /// Used to set the fab alignment.
  final Alignment? fabAlignment;

  /// The color of the menu background overlay.
  final Color? overlayColor;

  /// The opacity of the menu background overlay.
  /// Default: 0.8
  final double? overlayOpacity;

  /// Used to set the elevation of the fab button.
  /// Default: 1.0
  final double? elevation;

  /// Children buttons of the menu, from the lowest to the highest.
  final List<MenuItem> children;

  /// Animated FAB menu button.
  const FabMenu(
      {Key? key,
      this.fabIcon,
      this.fabBackgroundColor,
      this.fabAlignment,
      this.overlayColor,
      this.overlayOpacity,
      this.elevation,
      this.closeMenuButton,
      required this.children})
      : super(key: key);

  @override
  _FabMenuState createState() => _FabMenuState();
}

class _FabMenuState extends State<FabMenu> {
  @override
  Widget build(BuildContext context) {
    return Align(
      alignment: widget.fabAlignment ?? Alignment.bottomCenter,
      child: FloatingActionButton(
        heroTag: 'main-btn-tag',
        backgroundColor: widget.fabBackgroundColor ?? Colors.black,
        elevation: widget.elevation ?? 1.0,
        onPressed: () {
          Navigator.of(context).push(
            HeroDialogRoute(
              builder: (context) {
                return MainMenu(
                  closeMenuButton: widget.closeMenuButton ??
                      const Icon(
                        Icons.arrow_back,
                        color: Colors.white,
                      ),
                  backgroundColor: widget.overlayColor ?? Colors.black,
                  overlayOpacity: widget.overlayOpacity ?? 0.8,
                  children: widget.children,
                );
              },
            ),
          );
        },
        child: widget.fabIcon ?? const Icon(Icons.more_horiz),
      ),
    );
  }
}

// Full screen main menu
class MainMenu extends StatefulWidget {
  /// Used to set the menu close Icon.
  final Widget closeMenuButton;

  /// Used to get the background color of the menu list.
  final Color backgroundColor;

  /// Used to get the opacity of the background color of the menu list.
  final double overlayOpacity;

  /// Children buttons of the menu, from the lowest to the highest.
  final List<MenuItem> children;

  const MainMenu(
      {Key? key,
      required this.backgroundColor,
      required this.overlayOpacity,
      required this.closeMenuButton,
      required this.children})
      : super(key: key);

  @override
  State<MainMenu> createState() => _MainMenuState();
}

class _MainMenuState extends State<MainMenu> {
  @override
  Widget build(BuildContext context) {
    List<Widget> _getChildrenList() {
      return widget.children
          .map((MenuItem child) {
            int index = widget.children.indexOf(child);
            return Column(
              children: [
                ItemChild(
                  index: index,
                  title: child.title,
                  style: child.style,
                  onTap: child.onTap,
                ),
                const SizedBox(
                  height: 20,
                ),
              ],
            );
          })
          .toList()
          .reversed
          .toList();
    }

    var fabChildren = _getChildrenList();

    return Hero(
      tag: 'main-btn-tag',
      child: Material(
        color: widget.backgroundColor.withOpacity(widget.overlayOpacity),
        child: SingleChildScrollView(
          physics: const BouncingScrollPhysics(),
          child: SizedBox(
            height: MediaQuery.of(context).size.height,
            width: MediaQuery.of(context).size.width,
            child: Padding(
              padding: const EdgeInsets.all(32.0),
              child: Column(
                children: [
                  Expanded(
                    flex: 3,
                    child: ListView(
                      physics: const BouncingScrollPhysics(),
                      children: List.from(fabChildren),
                      reverse: true,
                    ),
                  ),
                  const SizedBox(
                    height: 20,
                  ),
                  Expanded(
                    flex: 1,
                    child: GestureDetector(
                      onTap: () {
                        Navigator.pop(context);
                      },
                      child: Container(
                        height: 50,
                        width: 100,
                        color: Colors.transparent,
                        child: widget.closeMenuButton,
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}


lib/src/hero_dialog_route.dart
import 'dart:ui';

import 'package:flutter/material.dart';

/// {@template hero_dialog_route}
/// Custom [PageRoute] that creates an overlay dialog (popup effect).
///
/// Best used with a [Hero] animation.
/// {@endtemplate}
class HeroDialogRoute<T> extends PageRoute<T> {
  /// {@macro hero_dialog_route}
  HeroDialogRoute({
    required WidgetBuilder builder,
    RouteSettings? settings,
    bool fullscreenDialog = false,
  })  : _builder = builder,
        super(settings: settings, fullscreenDialog: fullscreenDialog);

  final WidgetBuilder _builder;

  @override
  bool get opaque => false;

  @override
  bool get barrierDismissible => true;

  @override
  Duration get transitionDuration => const Duration(milliseconds: 500);

  @override
  bool get maintainState => true;

  @override
  Color get barrierColor => Colors.black54;

  @override
  Widget buildTransitions(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    return child;
  }

  @override
  Widget buildPage(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation) {
    return _builder(context);
  }

  @override
  String get barrierLabel => 'Popup dialog open';
}


exemple/lib/main.dart
import 'package:animated_fab_button_menu/animated_fab_button_menu.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Animated FAB Menu',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButton: FabMenu(
          fabBackgroundColor: Colors.amber,
          elevation: 2.0,
          fabAlignment: Alignment.bottomCenter,
          fabIcon: const Icon(Icons.more_horiz),
          closeMenuButton: const Icon(
            Icons.arrow_back,
            color: Colors.white,
          ),
          overlayOpacity: 0.5,
          overlayColor: Colors.amber,
          children: [
            MenuItem(
              title: 'Home',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Profile',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Messages',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Feedback',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Settings',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Share',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
            MenuItem(
              title: 'Get in touch!',
              onTap: () {},
              style: const TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold),
            ),
          ]),
      body: const Center(
        child: Text('Flutter animated FAB Menu'),
      ),
    );
  }
}
